src/client/async.rs

```diff
@@ -639,34 +640,26 @@ impl Client {
     /// the thread to exit.
     ///
     /// If the HTTP/2 connection cannot be initialized returns `None`.
-    pub fn with_connector<C, S>(connector: C) -> Option<Client>
-        where C: HttpConnect<Stream = S>,
-              S: TransportStream + Send + 'static
+    pub fn with_connector<C, S, E>(connector: C) -> Result<Client, ClientConnectError<E>>
+        where C: HttpConnect<Stream = S, Err = E>,
+              S: TransportStream + Send + 'static,
+              E: HttpConnectError + 'static
     {
         // Use the provided connector to establish a network connection...
-        let client_stream = match connector.connect().ok() {
-            Some(cs) => cs,
-            None => return None,
-        };
+        let client_stream = try!(connector.connect());
+
         // Keep a socket handle in order to shut it down once the service stops. This is required
         // because if the service decides to stop (due to all clients disconnecting) while the
         // socket is still open and the read thread waiting, it can happen that the read thread
         // (and as such the socket itself) ends up waiting indefinitely (or well, until the server
         // decides to close it), effectively leaking the socket and thread.
-        let mut sck = match client_stream.0.try_split() {
-            Ok(sck) => sck,
-            Err(_) => return None,
-        };
+        let mut sck = try!(client_stream.0.try_split());
+
+        let service = ClientService::new(client_stream);
 
-        let service = match ClientService::new(client_stream) {
-            Some(service) => service,
-            None => return None,
-        };
         let Service(mut service, rx, mut recv_frame, mut send_frame) = service;
 
-        if let Err(_) = rx.send(WorkItem::NewClient) {
-            return None;
-        }
+        rx.send(WorkItem::NewClient).expect("service is in scope and holds the receiver");
```

Comment:
The consumer of this API would need to handle a SendError which won't ever really come up. What do you think about just incrementing the client count manually after `ClientService::new()` instead of sending a message through the channel?